src/kubernetes_cluster/proof/cluster.rs

```diff
@@ -125,6 +125,7 @@ pub open spec fn desired_state_is(cr: K) -> StatePred<Self>
         &&& s.resource_key_exists(cr.object_ref())
         &&& K::from_dynamic_object(s.resource_obj_of(cr.object_ref())).is_Ok()
         &&& K::from_dynamic_object(s.resource_obj_of(cr.object_ref())).get_Ok_0().spec() == cr.spec()
+        &&& K::from_dynamic_object(s.resource_obj_of(cr.object_ref())).get_Ok_0().metadata().uid == cr.metadata().uid
```

Comment:
Actually, we don't have to change the desired_state_is predicate -- the fact that the cr always exists implies that its uid never changes. Could you leave a TODO here that later we will revert this predicate back?

Reply:
I don't understand. Do you mean `[]cr_key_exists |= K::from_dynamic_object(s.resource_obj_of(cr.object_ref())).get_Ok_0().metadata().uid == cr.metadata().uid`?

The `cr` here can be an arbitrary object, `[]cr_key_exists` only shows that the uid of the cr object stored in etcd never changes, not the parameter `cr` here.

Reply:
No, we don't want to say the uid is the same as `cr.metadata().uid`.
I was saying that we should treat uid similarly to the rest_id in the proof.
Basically, we no longer assume `K::from_dynamic_object(s.resource_obj_of(cr.object_ref())).get_Ok_0().metadata().uid == cr.metadata().uid`. Instead, you carry an `uid` as an argument (like the `rest_id`) in whatever lemma that needs the uid, with an extra precondition saying that `[]K::from_dynamic_object(s.resource_obj_of(cr.object_ref())).get_Ok_0().metadata().uid == uid`.
Then, you can prove that whatever `uid`, if initially we have `K::from_dynamic_object(s.resource_obj_of(cr.object_ref())).get_Ok_0().metadata().uid == uid`, then it is always true that `K::from_dynamic_object(s.resource_obj_of(cr.object_ref())).get_Ok_0().metadata().uid == uid` (that is, uid remains unchanaged).
Finally, you can use `leads_to_exists_intro` to eliminate the `uid`. Just like https://github.com/vmware-research/verifiable-controllers/blob/main/src/controller_examples/zookeeper_controller/proof/liveness/liveness_property.rs#L208-L238

Reply:
btw, I just realized that we should later revise `pub open spec fn desired_state_is(cr: K) -> StatePred<Self>` to `pub open spec fn desired_state_is(key: ObjectRef, spec: K::Spec) -> StatePred<Self>`. We only need the key and the spec here.

Reply:
btw2, can we revise `K::from_dynamic_object(s.resource_obj_of(cr.object_ref())).get_Ok_0().metadata().uid` to `s.resource_obj_of(cr.object_ref()).metadata.uid`?

Reply:
> btw, I just realized that we should later revise `pub open spec fn desired_state_is(cr: K) -> StatePred<Self>` to `pub open spec fn desired_state_is(key: ObjectRef, spec: K::Spec) -> StatePred<Self>`. We only need the key and the spec here. I didn't do that because associated type was not there.

Well, this could involve A LOT OF changes... Let's table it and leave a todo there maybe.

Reply:
> No, we don't want to say the uid is the same as cr.metadata().uid

This doesn't make sense to me. If the uid is not same as cr.metadata().uid, why should we care about that cr? I just think that we should only prove the liveness property about cr that always exists, so I think `desired_state_is` should say that that cr exists in ectd except the status thing you mentioned could be different.

Reply:
> why should we care about that cr?

I mean, we don't need to say that in the `desired_state_is()` (we want to minimize our assumption, if part of the predicate can be derived by the other parts), but I think it does no harm to keep it for now

Reply:
nvm, let's also table this for now

Reply:
Minimizing the assumption usually means getting a stronger property. But if strengthening the property is not useful, why should we minimize that?

Reply:
I am confused: if the assumption can be simpler (weaker) then why is it a bad thing? Or are you worried that weakening the assumption (for just one line) means many changes in the proof, so it is not worthwhile -- that I understand and agree.